No way to enforce @Test(singleThreaded = true) when test defined in base class

### TestNG Version

6.10, 7.3.0

### Expected behavior

Methods executed serially


### Actual behavior

Tests executed parallel

### Is the issue reproductible on runner?

- [ ] Shell
- [ ] Maven
- [ ] Gradle
- [ ] Ant
- [ ] Eclipse
- [x] IntelliJ
- [ ] NetBeans


### Test case sample

```java
@Test(singleThreaded = true)
public abstract class BaseTestClass
{
    private final AtomicInteger currentTests = new AtomicInteger();

    protected void test()
    {
        int currentTests = this.currentTests.incrementAndGet();
        try {
            assertEquals(currentTests, 1);
            MILLISECONDS.sleep(10);
        }
        catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
        finally {
            this.currentTests.decrementAndGet();
        }
    }

    @Test
    public void test1()
    {
        test();
    }

    @Test
    public void test2()
    {
        test();
    }

    @Test
    public void test3()
    {
        test();
    }

    @Test
    public void test4()
    {
        test();
    }
}
```

```java
@Test(singleThreaded = true)
public class TestClass
        extends BaseTestClass
{
}
```

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd">
<suite name="A Suite" parallel="methods" thread-count="4">
    <test name="A test" parallel="methods" thread-count="4">
        <packages>
            <package name="findepi.test.ng" />
        </packages>
    </test>
</suite>
```

> Why don't you attach a maven sample project that I can execute from the command line to reproduce the problem ?

https://github.com/findepi/testng-2361

on my machine:

```
$ mvn --version
Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Maven home: /usr/local/Cellar/maven/3.6.3_1/libexec
Java version: 11.0.7, vendor: Azul Systems, Inc., runtime: /Library/Java/JavaVirtualMachines/zulu-11.jdk/Contents/Home
Default locale: en_PL, platform encoding: UTF-8
OS name: "mac os x", version: "10.15.4", arch: "x86_64", family: "mac"
```

```
$ mvn clean test
[....]
[ERROR] Failures:
[ERROR]   TestClass>BaseTestClass.test2:52->BaseTestClass.test:32 expected [1] but found [2]
[ERROR]   TestClass>BaseTestClass.test3:58->BaseTestClass.test:32 expected [1] but found [2]
[ERROR]   TestClass>BaseTestClass.test4:64->BaseTestClass.test:32 expected [1] but found [2]
[ERROR] testng2361.TestClassWithFactory.test1(testng2361.TestClassWithFactory)
[INFO]   Run 1: PASS
[ERROR]   Run 2: TestClassWithFactory>BaseTestClass.test1:46->BaseTestClass.test:32 expected [1] but found [2]
[INFO]
[ERROR] testng2361.TestClassWithFactory.test2(testng2361.TestClassWithFactory)
[INFO]   Run 1: PASS
[ERROR]   Run 2: TestClassWithFactory>BaseTestClass.test2:52->BaseTestClass.test:32 expected [1] but found [2]
[INFO]
[ERROR] testng2361.TestClassWithFactory.test4(testng2361.TestClassWithFactory)
[ERROR]   Run 1: TestClassWithFactory>BaseTestClass.test4:64->BaseTestClass.test:32 expected [1] but found [2]
[ERROR]   Run 2: TestClassWithFactory>BaseTestClass.test4:64->BaseTestClass.test:32 expected [1] but found [2]
[INFO]
[INFO]
[ERROR] Tests run: 9, Failures: 6, Errors: 0, Skipped: 0
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[...]
```

--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -1,4 +1,5 @@
 Current
+Fixed: GITHUB-2361: No way to enforce @Test(singleThreaded = true) when test defined in base class (Krishnan Mahadevan)
 Fixed: GITHUB-2343: Injectors are not reused when they share the same set of modules (Krishnan Mahadevan)
 Fixed: GITHUB-2346: ITestResult attributes are null when retrieved by Listener onTestStart if test fails at BeforeMethod (Krishnan Mahadevan)
 Fixed: GITHUB-2357: TestNG 7.3.0 transitive dependencies

--- a/src/main/java/org/testng/internal/ClassBasedParallelWorker.java
+++ b/src/main/java/org/testng/internal/ClassBasedParallelWorker.java
@@ -17,11 +17,7 @@ import java.util.stream.Collectors;
 
 class ClassBasedParallelWorker extends AbstractParallelWorker {
 
-  @Override
-  public List<IWorker<ITestNGMethod>> createWorkers(Arguments arguments) {
-    List<IWorker<ITestNGMethod>> result = Lists.newArrayList();
-    // Methods that belong to classes with a sequential=true or parallel=classes
-    // attribute must all be run in the same worker
+  private static Set<Class<?>> gatherClassesThatShouldRunSequentially(Arguments arguments) {
     Set<Class<?>> sequentialClasses = Sets.newHashSet();
     for (ITestNGMethod m : arguments.getMethods()) {
       Class<?> cls = m.getRealClass();
@@ -33,6 +29,15 @@ class ClassBasedParallelWorker extends AbstractParallelWorker {
         sequentialClasses.add(cls);
       }
     }
+    return sequentialClasses;
+  }
+
+  @Override
+  public List<IWorker<ITestNGMethod>> createWorkers(Arguments arguments) {
+    List<IWorker<ITestNGMethod>> result = Lists.newArrayList();
+    // Methods that belong to classes with a sequential=true or parallel=classes
+    // attribute must all be run in the same worker
+    Set<Class<?>> sequentialClasses = gatherClassesThatShouldRunSequentially(arguments);
 
     List<IMethodInstance> methodInstances = Lists.newArrayList();
     for (ITestNGMethod tm : arguments.getMethods()) {
@@ -49,7 +54,7 @@ class ClassBasedParallelWorker extends AbstractParallelWorker {
         params = getParameters(im);
         prevClass = c;
       }
-      if (sequentialClasses.contains(c)) {
+      if (shouldRunSequentially(c, sequentialClasses)) {
         if (!processedClasses.contains(c)) {
           processedClasses.add(c);
           // Sequential class: all methods in one worker
@@ -65,6 +70,11 @@ class ClassBasedParallelWorker extends AbstractParallelWorker {
     }
 
     return result;
+  }
+
+  private static boolean shouldRunSequentially(Class<?> c, Set<Class<?>> sequentialClasses) {
+    return sequentialClasses.contains(c) ||
+        sequentialClasses.stream().anyMatch(each -> each.isAssignableFrom(c));
   }
 
   private static List<IMethodInstance> findClasses(

--- a/src/test/java/test/thread/issue2361/AnotherChildClassExample.java
+++ b/src/test/java/test/thread/issue2361/AnotherChildClassExample.java
@@ -1,0 +1,8 @@
+package test.thread.issue2361;
+
+import org.testng.annotations.Test;
+
+@Test(singleThreaded = true)
+public class AnotherChildClassExample extends BaseTestClassExample {
+
+}

--- a/src/test/java/test/thread/issue2361/BaseTestClassExample.java
+++ b/src/test/java/test/thread/issue2361/BaseTestClassExample.java
@@ -1,0 +1,45 @@
+package test.thread.issue2361;
+
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static org.testng.Assert.assertEquals;
+
+import java.util.concurrent.atomic.AtomicInteger;
+import org.testng.annotations.Test;
+
+@Test(singleThreaded = true)
+public class BaseTestClassExample {
+
+  private final AtomicInteger currentTests = new AtomicInteger();
+
+  protected void test() {
+    int currentTests = this.currentTests.incrementAndGet();
+    try {
+      assertEquals(currentTests, 1);
+      MILLISECONDS.sleep(10);
+    } catch (InterruptedException e) {
+      throw new RuntimeException(e);
+    } finally {
+      this.currentTests.decrementAndGet();
+    }
+  }
+
+  @Test
+  public void test1() {
+    test();
+  }
+
+  @Test
+  public void test2() {
+    test();
+  }
+
+  @Test
+  public void test3() {
+    test();
+  }
+
+  @Test
+  public void test4() {
+    test();
+  }
+}

--- a/src/test/java/test/thread/issue2361/ChildClassExample.java
+++ b/src/test/java/test/thread/issue2361/ChildClassExample.java
@@ -1,0 +1,8 @@
+package test.thread.issue2361;
+
+import org.testng.annotations.Test;
+
+@Test(singleThreaded = true)
+public class ChildClassExample extends BaseTestClassExample {
+
+}

--- a/src/test/java/test/thread/issue2361/FactorySample.java
+++ b/src/test/java/test/thread/issue2361/FactorySample.java
@@ -1,0 +1,15 @@
+package test.thread.issue2361;
+
+import org.testng.annotations.Factory;
+
+public class FactorySample {
+
+  @Factory
+  public static Object[] newInstances() {
+    return new Object[] {
+        new ChildClassExample(),
+        new AnotherChildClassExample()
+    };
+  }
+
+}

--- a/src/test/java/test/thread/issue2361/IssueTest.java
+++ b/src/test/java/test/thread/issue2361/IssueTest.java
@@ -1,0 +1,32 @@
+package test.thread.issue2361;
+
+import org.assertj.core.api.Assertions;
+import org.testng.TestNG;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+import org.testng.xml.XmlSuite;
+import org.testng.xml.XmlSuite.ParallelMode;
+import org.testng.xml.XmlTest;
+import test.SimpleBaseTest;
+
+public class IssueTest extends SimpleBaseTest {
+
+  @Test(dataProvider = "dp")
+  public void ensureClassLevelSingleThreadedNatureGetsHonoured(Class<?> cls) {
+    XmlSuite suite = createXmlSuite("Sample_Suite");
+    suite.setParallel(ParallelMode.METHODS);
+    XmlTest xmlTest = createXmlTest(suite, "Sample_Test", cls);
+    xmlTest.setParallel(ParallelMode.METHODS);
+    TestNG testng = create(suite);
+    testng.run();
+    Assertions.assertThat(testng.getStatus()).isEqualTo(0);
+  }
+
+  @DataProvider(name = "dp")
+  public Object[][] getTestData() {
+    return new Object[][] {
+        {ChildClassExample.class},
+        {FactorySample.class}
+    };
+  }
+}

--- a/src/test/resources/testng.xml
+++ b/src/test/resources/testng.xml
@@ -99,6 +99,7 @@
       <class name="test.mannotation.MAnnotation2SampleTest" />
       <class name="test.mannotation.issue1976.IssueTest" />
       <class name="test.thread.FactoryTest" />
+      <class name="test.thread.issue2361.IssueTest"/>
       <class name="test.thread.SequentialTest" />
       <class name="test.thread.ParallelTestTest" />
       <class name="test.thread.DataProviderThreadPoolSizeTest" />

